AMDGPU: Don't avoid clamp of bit shift in BFE pattern

changpeng opened the pull request:
Enable pattern matching from "x<<32-y>>32-y" to "bfe x, 0, y" when we know y is in [0,31].
  This is the follow-up for the PR: https://github.com/llvm/llvm-project/pull/114279 to fix the issue: https://github.com/llvm/llvm-project/issues/114282

changpeng committed:
AMDGPU: Don't avoid clamp of bit shift in BFE pattern

  Enable pattern matching from "x<<32-y>>32-y" to "bfe x, 0, y"
when we know y is in [0,31].
  This is the follow-up for the PR: https://github.com/llvm/llvm-project/pull/114279
to fix the issue: https://github.com/llvm/llvm-project/issues/114282

rampitec commented on llvm/lib/Target/AMDGPU/SIInstructions.td:
```diff
@@ -3553,6 +3553,23 @@ def : AMDGPUPat <
 >;
 
+def uint5Bits : PatLeaf<(i32 VGPR_32:$width), [{
+  return CurDAG->computeKnownBits(SDValue(N, 0)).countMinLeadingZeros() >= 27;
```
`countMaxTrailingOnes() <= 5` would be easier to read.

rampitec reviewed (approved):
LGTM with a nit.

changpeng committed:
AMDGPU: Don't avoid clamp of bit shift in BFE pattern

  Use "countMaxTrailingOnes() <= 5" instead of "countMinLeadingZeros() >= 27"
in uint5Bits PatLeaf definition.

arsenm commented on llvm/test/CodeGen/AMDGPU/extract-lowbits.ll:
```diff
@@ -150,22 +150,14 @@ define i32 @bzhi32_c4_commutative(i32 %val, i32 %numlowbits) nounwind {
-; VI-NEXT:    v_lshrrev_b32_e32 v0, v1, v0
-; VI-NEXT:    s_setpc_b64 s[30:31]
-  %numhighbits = sub i32 32, %numlowbits
+; GCN-LABEL: bzhi32_d0:
```
This isn't canonical IR. We shouldn't be burning compile time optimizing non-canonical IR. Can you add a copy of this test with the instcombine output, and revert this if you get the same output

changpeng replied:
Instcombine does not change the test source, and I think extracting the low bits are frequently used in the programs to restrict the range of a value.

arsenm replied:
Instcombine does change this: 


```
define i32 @bzhi32_d0(i32 %val, i32 %numlowbits) #0 {
  %numlow5bits = and i32 %numlowbits, 31
  %numhighbits = sub nuw nsw i32 32, %numlow5bits
  %1 = lshr i32 -1, %numhighbits
  %masked = and i32 %1, %val
  ret i32 %masked
}
```

Codegenning this does produce BFE already (except we fail to optimize out the clamp, which is a separate issue)

jayfoad commented on llvm/lib/Target/AMDGPU/SIInstructions.td:
```diff
@@ -3553,6 +3553,23 @@ def : AMDGPUPat <
 >;
 
+def uint5Bits : PatLeaf<(i32 VGPR_32:$width), [{
+  return CurDAG->computeKnownBits(SDValue(N, 0)).countMaxTrailingOnes() <= 5;
```
No, this is completely wrong!!! E.g. if bit 0 is known to be 0 then `countMaxTrailingOnes` will return 0 and this test will pass.

rampitec replied:
Ugh, my bad, sorry.

changpeng replied:
Thanks so much.  Also countMaxActiveBits is more close to what we are going to define

jayfoad commented on llvm/lib/Target/AMDGPU/SIInstructions.td:
```diff
@@ -3553,6 +3553,23 @@ def : AMDGPUPat <
 >;
 
+def uint5Bits : PatLeaf<(i32 VGPR_32:$width), [{
+  return CurDAG->computeKnownBits(SDValue(N, 0)).countMaxTrailingOnes() <= 5;
```
```suggestion
  return CurDAG->computeKnownBits(SDValue(N, 0)).countMaxActiveBits() <= 5;
```
Please also add a test that shows why countMaxTrailingOnes was wrong.

changpeng replied:
Thanks, Done in https://github.com/llvm/llvm-project/pull/115543
